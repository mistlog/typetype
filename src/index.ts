export * from "./parser";
export * from "./generator";
export * from "./core";