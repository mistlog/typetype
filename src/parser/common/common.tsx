import { ReactPeg } from "react-peg";

// white space
export function _() {
    return (
        <repeat type="*">
            <set> \t\n\r</set>
        </repeat>
    );
}

export const Text = (content: string) => {
    return (
        <list>
            <_ /><text>{content}</text><_ />
        </list>
    )
}