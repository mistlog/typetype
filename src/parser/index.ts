export * from "./common";
export * from "./basic";
export * from "./function";
export * from "./expression";
export * from "./statement";
export * from "./declaration";
export * from "./start";