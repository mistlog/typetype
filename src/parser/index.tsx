export * from "./common";
export * from "./basic";
export * from "./expression";
export * from "./statement";
export * from "./declaration";
export * from "./start";