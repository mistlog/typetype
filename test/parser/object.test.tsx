import { ReactPeg } from "react-peg";
import { TypeObjectProperty, ObjectTypeLiteral } from "../../src";
import { saveAST } from "../common";

describe("parser: object", () => {

    test("ObjectType: call signature", () => {
        const parser = ReactPeg.render(<ObjectTypeLiteral />);
        const ast = parser.parse(`
            {
                new (someArg: number): boolean,
                (a: number, b: number): number
            }
        `);
        saveAST(ast, "ObjectType-CallSignature.json");
        expect(ast).toMatchSnapshot();
    
        //
        expect(parser.parse(`{}`)).toMatchSnapshot();
    })

    test("ObjectType: modifier", () => {
            const parser = ReactPeg.render(<ObjectTypeLiteral />);
            const ast = parser.parse(`
                {
                    readonly a ?: number
                }
            `);

            saveAST(ast, "ObjectType-Modifier.json");
            expect(ast).toMatchSnapshot();
    })

    test("ObjectType", () => {
        const parser = ReactPeg.render(<ObjectTypeLiteral />);
        const ast = parser.parse(`
            {
                a: number,
                b: string,
                c: 1,
                d: "abc"
            }
        `);
        saveAST(ast, "ObjectType.json");
        expect(ast).toMatchSnapshot();
    
        //
        expect(parser.parse(`{}`)).toMatchSnapshot();
    })

    test("ObjectType: whitespace in key", () => {
        const parser = ReactPeg.render(<ObjectTypeLiteral />);
        const ast = parser.parse(`
            {
                "temp value": number
            }
        `);
        saveAST(ast, "ObjectType-Key-WhiteSpace.json");
        expect(ast).toMatchSnapshot();
    })
    
    test("ObjectType: shorthand", () => {
        const parser = ReactPeg.render(<ObjectTypeLiteral />);
        const ast = parser.parse(`
            {
                hello
            }
        `);
        saveAST(ast, "ObjectType-Shorthand.json");
        expect(ast).toMatchSnapshot();
    })
    
    test("ObjectType: spread", () => {
        const parser = ReactPeg.render(<ObjectTypeLiteral />);
        const ast = parser.parse(`
            {
                hello,
                ...obj
            }
        `);
        saveAST(ast, "ObjectType-Spread.json");
        expect(ast).toMatchSnapshot();
    })

    test("TypeObjectProperty", () => {
        const parser = ReactPeg.render(<TypeObjectProperty />);
        expect(parser.parse(`a: number`)).toMatchSnapshot();
        expect(parser.parse(`b: string`)).toMatchSnapshot();
        expect(parser.parse(`c: 1`)).toMatchSnapshot();
        expect(parser.parse(`d: "abc"`)).toMatchSnapshot();
    })
})