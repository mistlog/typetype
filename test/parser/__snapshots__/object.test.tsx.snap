// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser: object ObjectType 1`] = `
Object {
  "kind": "ObjectTypeLiteral",
  "props": Array [
    Object {
      "kind": "TypeObjectProperty",
      "name": Object {
        "kind": "Identifier",
        "name": "a",
      },
      "optional": false,
      "readonly": false,
      "value": Object {
        "kind": "NumberType",
        "value": "number",
      },
    },
    Object {
      "kind": "TypeObjectProperty",
      "name": Object {
        "kind": "Identifier",
        "name": "b",
      },
      "optional": false,
      "readonly": false,
      "value": Object {
        "kind": "StringType",
        "value": "string",
      },
    },
    Object {
      "kind": "TypeObjectProperty",
      "name": Object {
        "kind": "Identifier",
        "name": "c",
      },
      "optional": false,
      "readonly": false,
      "value": Object {
        "kind": "NumberTypeLiteral",
        "value": 1,
      },
    },
    Object {
      "kind": "TypeObjectProperty",
      "name": Object {
        "kind": "Identifier",
        "name": "d",
      },
      "optional": false,
      "readonly": false,
      "value": Object {
        "kind": "StringTypeLiteral",
        "value": "abc",
      },
    },
  ],
}
`;

exports[`parser: object ObjectType 2`] = `
Object {
  "kind": "ObjectTypeLiteral",
  "props": Array [],
}
`;

exports[`parser: object ObjectType: call signature 1`] = `
Object {
  "kind": "ObjectTypeLiteral",
  "props": Array [
    Object {
      "kind": "TypeObjectProperty",
      "name": Object {
        "kind": "ConstructSignature",
        "params": Array [
          Object {
            "kind": "FunctionTypeParam",
            "name": Object {
              "kind": "Identifier",
              "name": "someArg",
            },
            "type": Object {
              "kind": "NumberType",
              "value": "number",
            },
          },
        ],
      },
      "optional": false,
      "readonly": false,
      "value": Object {
        "kind": "TypeReference",
        "typeName": Object {
          "kind": "Identifier",
          "name": "boolean",
        },
      },
    },
    Object {
      "kind": "TypeObjectProperty",
      "name": Object {
        "kind": "CallSignature",
        "params": Array [
          Object {
            "kind": "FunctionTypeParam",
            "name": Object {
              "kind": "Identifier",
              "name": "a",
            },
            "type": Object {
              "kind": "NumberType",
              "value": "number",
            },
          },
          Object {
            "kind": "FunctionTypeParam",
            "name": Object {
              "kind": "Identifier",
              "name": "b",
            },
            "type": Object {
              "kind": "NumberType",
              "value": "number",
            },
          },
        ],
      },
      "optional": false,
      "readonly": false,
      "value": Object {
        "kind": "NumberType",
        "value": "number",
      },
    },
  ],
}
`;

exports[`parser: object ObjectType: call signature 2`] = `
Object {
  "kind": "ObjectTypeLiteral",
  "props": Array [],
}
`;

exports[`parser: object ObjectType: modifier 1`] = `
Object {
  "kind": "ObjectTypeLiteral",
  "props": Array [
    Object {
      "kind": "TypeObjectProperty",
      "name": Object {
        "kind": "Identifier",
        "name": "a",
      },
      "optional": true,
      "readonly": true,
      "value": Object {
        "kind": "NumberType",
        "value": "number",
      },
    },
  ],
}
`;

exports[`parser: object ObjectType: shorthand 1`] = `
Object {
  "kind": "ObjectTypeLiteral",
  "props": Array [
    Object {
      "kind": "TypeObjectProperty",
      "name": Object {
        "kind": "Identifier",
        "name": "hello",
      },
      "optional": false,
      "readonly": false,
      "value": Object {
        "kind": "TypeReference",
        "typeName": Object {
          "kind": "Identifier",
          "name": "hello",
        },
      },
    },
  ],
}
`;

exports[`parser: object ObjectType: spread 1`] = `
Object {
  "kind": "ObjectTypeLiteral",
  "props": Array [
    Object {
      "kind": "TypeObjectProperty",
      "name": Object {
        "kind": "Identifier",
        "name": "hello",
      },
      "optional": false,
      "readonly": false,
      "value": Object {
        "kind": "TypeReference",
        "typeName": Object {
          "kind": "Identifier",
          "name": "hello",
        },
      },
    },
    Object {
      "kind": "TypeSpreadProperty",
      "param": Object {
        "kind": "TypeReference",
        "typeName": Object {
          "kind": "Identifier",
          "name": "obj",
        },
      },
    },
  ],
}
`;

exports[`parser: object ObjectType: whitespace in key 1`] = `
Object {
  "kind": "ObjectTypeLiteral",
  "props": Array [
    Object {
      "kind": "TypeObjectProperty",
      "name": Object {
        "kind": "StringLiteral",
        "value": "temp value",
      },
      "optional": false,
      "readonly": false,
      "value": Object {
        "kind": "NumberType",
        "value": "number",
      },
    },
  ],
}
`;

exports[`parser: object TypeObjectProperty 1`] = `
Object {
  "kind": "TypeObjectProperty",
  "name": Object {
    "kind": "Identifier",
    "name": "a",
  },
  "optional": false,
  "readonly": false,
  "value": Object {
    "kind": "NumberType",
    "value": "number",
  },
}
`;

exports[`parser: object TypeObjectProperty 2`] = `
Object {
  "kind": "TypeObjectProperty",
  "name": Object {
    "kind": "Identifier",
    "name": "b",
  },
  "optional": false,
  "readonly": false,
  "value": Object {
    "kind": "StringType",
    "value": "string",
  },
}
`;

exports[`parser: object TypeObjectProperty 3`] = `
Object {
  "kind": "TypeObjectProperty",
  "name": Object {
    "kind": "Identifier",
    "name": "c",
  },
  "optional": false,
  "readonly": false,
  "value": Object {
    "kind": "NumberTypeLiteral",
    "value": 1,
  },
}
`;

exports[`parser: object TypeObjectProperty 4`] = `
Object {
  "kind": "TypeObjectProperty",
  "name": Object {
    "kind": "Identifier",
    "name": "d",
  },
  "optional": false,
  "readonly": false,
  "value": Object {
    "kind": "StringTypeLiteral",
    "value": "abc",
  },
}
`;
